{"nbformat":4,"nbformat_minor":0,"metadata":{"colab":{"provenance":[{"file_id":"https://github.com/daveminh/Chem456-2022F/blob/main/labs/03-colab_fold.ipynb","timestamp":1662474132388},{"file_id":"https://github.com/sokrypton/ColabFold/blob/main/AlphaFold2.ipynb","timestamp":1660773744301}],"collapsed_sections":[]},"kernelspec":{"display_name":"Python 3 (ipykernel)","language":"python","name":"python3"},"language_info":{"codemirror_mode":{"name":"ipython","version":3},"file_extension":".py","mimetype":"text/x-python","name":"python","nbconvert_exporter":"python","pygments_lexer":"ipython3","version":"3.8.10"},"gpuClass":"standard","accelerator":"GPU"},"cells":[{"cell_type":"markdown","source":["This exercise is based on the ColabFold notebook. ColabFold predicts structures using AlphaFold2, a groundbreaking protein structure prediction method based on machine learning. It is based on a faster method for multiple sequence alignment than AlphaFold2.\n","\n","The exercise is intended to help you achieve the following learning objective:\n","* Predict protein structure based on the sequence of amino acids. Express confidence in the quality of a structure prediction.\n","\n","When you are done with this exercise, save it under `Chem456-2022F/exercises` on Google Drive. It will be graded as satisfactory or unsatisfactory based on correctly completing the sections after # -->."],"metadata":{"id":"lbX5Ktsq1Z4o"}},{"cell_type":"markdown","source":["# Overview\n","\n","The western corn rootworm (WCR) is an important agricultural pest that has evolved the ability to tolerate the plant defense toxin DIMBOA. Nick Miller (Biology) and I are interested in identifying WCR enzymes that metabolize DIMBOA. We hypothesize that detoxification is performed by cytochrome P450s, of which WCR has 85. In this exercise, you will build a model of two WCR enzymes.\n","\n","In Part I, you will use ColabFold to model the structure of one enzyme without cofactors. You will also assess this model and compare it to a model from AlphaFold2, which is based on a slower multiple sequence alignment. In Part II, you will model the position of the heme cofactor based on a homologous crystal structure.\n","\n","You will find the protein sequences, AlphaFold2 structures, and your assigned proteins in this folder:\n","https://drive.google.com/drive/folders/1BLshPMhersj9C6WzPyPIdoP-iBKhC_XO?usp=sharing\n","\n","If you are working on this exercise independently (opposed to being in Chem 456), you may try the sequence of the WCR CYP4AA5:\n","`MDYYLGYILTVLLTIFRTLKNYMRTVILAWKLPGPTALPILGNVLILKNHTELIYLGNHAPNMYGKFFRVWCSILPAIFITEPRHLKQILSTNKNNEKSLFYRAFHNFIGEGLITNNGYKWKKNRKMIQPYFHMNILEKFVDLFTECSSRFVSKLEGEQTVKITPFINDCVLDILHNGVLGVPFDQDSPYRKGQLQAIERFVRPWLLFDSIFDRTNSANYEKKQKHNLHSYTEEVLNNRRQEIKKGTKGTKTCFLDMFIEIADNNSQFTEEDIINEIVTFMLAGQDSVGATIAFSMYYLAKYPEIQDKVLKEISGIDATQKISIKELNEMKYLEQVVKETLRLTPAVPIITRVLTEDVTLDETVFPNGTIIFISPFMTQRLPHVFPDPNTFNPDRFEESNLENIHPYAFLPFSLGPRNCIGYKFAMIEVKTVLFYVLKKFQLSPAKGKEELSLSYRATIRASGGIWVNLKHRKS`"],"metadata":{"id":"7X4V0Hlt5APb"}},{"cell_type":"markdown","metadata":{"id":"G4yBrceuFbf3"},"source":["<img src=\"https://raw.githubusercontent.com/sokrypton/ColabFold/main/.github/ColabFold_Marv_Logo_Small.png\" height=\"200\" align=\"right\" style=\"height:240px\">\n","\n","# Part I. ColabFold\n","\n","Easy to use protein structure and complex prediction using [AlphaFold2](https://www.nature.com/articles/s41586-021-03819-2) and [Alphafold2-multimer](https://www.biorxiv.org/content/10.1101/2021.10.04.463034v1). Sequence alignments/templates are generated through [MMseqs2](mmseqs.com) and [HHsearch](https://github.com/soedinglab/hh-suite). For more details, see <a href=\"#Instructions\">bottom</a> of the notebook, checkout the [ColabFold GitHub](https://github.com/sokrypton/ColabFold) and read the manuscript. Old versions: [v1.0](https://colab.research.google.com/github/sokrypton/ColabFold/blob/v1.0-alpha/AlphaFold2.ipynb), [v1.1](https://colab.research.google.com/github/sokrypton/ColabFold/blob/v1.1-premultimer/AlphaFold2.ipynb), [v1.2](https://colab.research.google.com/github/sokrypton/ColabFold/blob/v1.2.0/AlphaFold2.ipynb), [v1.3](https://colab.research.google.com/github/sokrypton/ColabFold/blob/v1.3.0/AlphaFold2.ipynb)\n","\n","[Mirdita M, SchÃ¼tze K, Moriwaki Y, Heo L, Ovchinnikov S, Steinegger M. ColabFold: Making protein folding accessible to all.\n","*Nature Methods*, 2022](https://www.nature.com/articles/s41592-022-01488-1) "]},{"cell_type":"code","metadata":{"id":"kOblAo-xetgx","cellView":"form"},"source":["#@title Input protein sequence(s), then hit `Runtime` -> `Run all`\n","from google.colab import files\n","import os.path\n","import re\n","import hashlib\n","import random\n","\n","def add_hash(x,y):\n","  return x+\"_\"+hashlib.sha1(y.encode()).hexdigest()[:5]\n","\n","query_sequence = 'MDYYLGYILTVLLTIFRTLKNYMRTVILAWKLPGPTALPILGNVLILKNHTELIYLGNHAPNMYGKFFRVWCSILPAIFITEPRHLKQILSTNKNNEKSLFYRAFHNFIGEGLITNNGYKWKKNRKMIQPYFHMNILEKFVDLFTECSSRFVSKLEGEQTVKITPFINDCVLDILHNGVLGVPFDQDSPYRKGQLQAIERFVRPWLLFDSIFDRTNSANYEKKQKHNLHSYTEEVLNNRRQEIKKGTKGTKTCFLDMFIEIADNNSQFTEEDIINEIVTFMLAGQDSVGATIAFSMYYLAKYPEIQDKVLKEISGIDATQKISIKELNEMKYLEQVVKETLRLTPAVPIITRVLTEDVTLDETVFPNGTIIFISPFMTQRLPHVFPDPNTFNPDRFEESNLENIHPYAFLPFSLGPRNCIGYKFAMIEVKTVLFYVLKKFQLSPAKGKEELSLSYRATIRASGGIWVNLKHRKS' #@param {type:\"string\"}\n","#@markdown  - Use `:` to specify inter-protein chainbreaks for **modeling complexes** (supports homo- and hetro-oligomers). For example **PI...SK:PI...SK** for a homodimer\n","\n","# remove whitespaces\n","query_sequence = \"\".join(query_sequence.split())\n","\n","jobname = 'CYP4AA5' #@param {type:\"string\"}\n","# remove whitespaces\n","basejobname = \"\".join(jobname.split())\n","basejobname = re.sub(r'\\W+', '', basejobname)\n","jobname = add_hash(basejobname, query_sequence)\n","while os.path.isfile(f\"{jobname}.csv\"):\n","  jobname = add_hash(basejobname, ''.join(random.sample(query_sequence,len(query_sequence))))\n","\n","with open(f\"{jobname}.csv\", \"w\") as text_file:\n","    text_file.write(f\"id,sequence\\n{jobname},{query_sequence}\")\n","\n","queries_path=f\"{jobname}.csv\"\n","\n","# number of models to use\n","use_amber = False #@param {type:\"boolean\"}\n","template_mode = \"pdb70\" #@param [\"none\", \"pdb70\",\"custom\"]\n","#@markdown - \"none\" = no template information is used, \"pdb70\" = detect templates in pdb70, \"custom\" - upload and search own templates (PDB or mmCIF format, see [notes below](#custom_templates))\n","\n","if template_mode == \"pdb70\":\n","  use_templates = True\n","  custom_template_path = None\n","elif template_mode == \"custom\":\n","  custom_template_path = f\"{jobname}_template\"\n","  os.mkdir(custom_template_path)\n","  uploaded = files.upload()\n","  use_templates = True\n","  for fn in uploaded.keys():\n","    os.rename(fn, f\"{jobname}_template/{fn}\")\n","else:\n","  custom_template_path = None\n","  use_templates = False\n"],"execution_count":null,"outputs":[]},{"cell_type":"code","source":["#@markdown ### MSA options (custom MSA upload, single sequence, pairing mode)\n","msa_mode = \"MMseqs2 (UniRef+Environmental)\" #@param [\"MMseqs2 (UniRef+Environmental)\", \"MMseqs2 (UniRef only)\",\"single_sequence\",\"custom\"]\n","pair_mode = \"unpaired+paired\" #@param [\"unpaired+paired\",\"paired\",\"unpaired\"] {type:\"string\"}\n","#@markdown - \"unpaired+paired\" = pair sequences from same species + unpaired MSA, \"unpaired\" = seperate MSA for each chain, \"paired\" - only use paired sequences.\n","\n","# decide which a3m to use\n","if msa_mode.startswith(\"MMseqs2\"):\n","  a3m_file = f\"{jobname}.a3m\"\n","elif msa_mode == \"custom\":\n","  a3m_file = f\"{jobname}.custom.a3m\"\n","  if not os.path.isfile(a3m_file):\n","    custom_msa_dict = files.upload()\n","    custom_msa = list(custom_msa_dict.keys())[0]\n","    header = 0\n","    import fileinput\n","    for line in fileinput.FileInput(custom_msa,inplace=1):\n","      if line.startswith(\">\"):\n","         header = header + 1\n","      if not line.rstrip():\n","        continue\n","      if line.startswith(\">\") == False and header == 1:\n","         query_sequence = line.rstrip()\n","      print(line, end='')\n","\n","    os.rename(custom_msa, a3m_file)\n","    queries_path=a3m_file\n","    print(f\"moving {custom_msa} to {a3m_file}\")\n","else:\n","  a3m_file = f\"{jobname}.single_sequence.a3m\"\n","  with open(a3m_file, \"w\") as text_file:\n","    text_file.write(\">1\\n%s\" % query_sequence)"],"metadata":{"cellView":"form","id":"C2_sh2uAonJH"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["#@markdown ### Advanced settings\n","model_type = \"auto\" #@param [\"auto\", \"AlphaFold2-ptm\", \"AlphaFold2-multimer-v1\", \"AlphaFold2-multimer-v2\"]\n","#@markdown - \"auto\" = protein structure prediction using \"AlphaFold2-ptm\" and complex prediction \"AlphaFold-multimer-v2\". For complexes \"AlphaFold-multimer-v[1,2]\" and \"AlphaFold-ptm\" can be used.\n","num_recycles = 3 #@param [1,3,6,12,24,48] {type:\"raw\"}\n","save_to_google_drive = True #@param {type:\"boolean\"}\n","\n","#@markdown -  if the save_to_google_drive option was selected, the result zip will be uploaded to your Google Drive\n","dpi = 200 #@param {type:\"integer\"}\n","#@markdown - set dpi for image resolution\n","\n","#@markdown Don't forget to hit `Runtime` -> `Run all` after updating the form.\n","\n","# We will use google.colab instead of pydrive, which was in the original ColabFold\n","# if save_to_google_drive:\n","#   from pydrive.drive import GoogleDrive\n","#   from pydrive.auth import GoogleAuth\n","#   from google.colab import auth\n","#   from oauth2client.client import GoogleCredentials\n","#   auth.authenticate_user()\n","#   gauth = GoogleAuth()\n","#   gauth.credentials = GoogleCredentials.get_application_default()\n","#   drive = GoogleDrive(gauth)\n","#   print(\"You are logged into Google Drive and are good to go!\")"],"metadata":{"cellView":"form","id":"ADDuaolKmjGW"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["#@title Install dependencies using python\n","#@markdown If necessary, mount Google Drive. Install py3Dmol for structural visualization and prody for structural alignment.\n","\n","from google.colab import drive\n","if not os.path.isdir('/content/drive'):\n","  drive.mount('/content/drive')\n","\n","try:\n","  import py3Dmol\n","except:\n","  !pip install py3Dmol\n","  import py3Dmol\n","  \n","try:\n","  import prody\n","except:\n","  !pip install -U prody\n","  import prody"],"metadata":{"cellView":"form","id":"HOITnmnYKr3M"},"execution_count":null,"outputs":[]},{"cell_type":"code","metadata":{"id":"iccGdbe_Pmt9","pycharm":{"name":"#%%\n"},"cellView":"form"},"source":["#@title Install dependencies using bash\n","#@markdown If you have already completed Part I, you can skip ahead to Part II without running this cell.\n","%%bash -s $use_amber $use_templates\n","\n","set -e\n","\n","USE_AMBER=$1\n","USE_TEMPLATES=$2\n","\n","if [ ! -f COLABFOLD_READY ]; then\n","  # install dependencies\n","  # We have to use \"--no-warn-conflicts\" because colab already has a lot preinstalled with requirements different to ours\n","  pip install -q --no-warn-conflicts \"colabfold[alphafold-minus-jax] @ git+https://github.com/sokrypton/ColabFold\"\n","  # high risk high gain\n","  pip install -q \"jax[cuda11_cudnn805]>=0.3.8,<0.4\" -f https://storage.googleapis.com/jax-releases/jax_releases.html\n","  touch COLABFOLD_READY\n","fi\n","\n","# setup conda\n","if [ ${USE_AMBER} == \"True\" ] || [ ${USE_TEMPLATES} == \"True\" ]; then\n","  if [ ! -f CONDA_READY ]; then\n","    wget -qnc https://repo.anaconda.com/miniconda/Miniconda3-latest-Linux-x86_64.sh\n","    bash Miniconda3-latest-Linux-x86_64.sh -bfp /usr/local 2>&1 1>/dev/null\n","    rm Miniconda3-latest-Linux-x86_64.sh\n","    touch CONDA_READY\n","  fi\n","fi\n","# setup template search\n","if [ ${USE_TEMPLATES} == \"True\" ] && [ ! -f HH_READY ]; then\n","  conda install -y -q -c conda-forge -c bioconda kalign2=2.04 hhsuite=3.3.0 python=3.7 2>&1 1>/dev/null\n","  touch HH_READY\n","fi\n","# setup openmm for amber refinement\n","if [ ${USE_AMBER} == \"True\" ] && [ ! -f AMBER_READY ]; then\n","  conda install -y -q -c conda-forge openmm=7.5.1 python=3.7 pdbfixer 2>&1 1>/dev/null\n","  touch AMBER_READY\n","fi"],"execution_count":null,"outputs":[]},{"cell_type":"code","metadata":{"id":"_sztQyz29DIC","cellView":"form"},"source":["#@title Run Prediction\n","\n","import sys\n","\n","from colabfold.download import download_alphafold_params, default_data_dir\n","from colabfold.utils import setup_logging\n","from colabfold.batch import get_queries, run, set_model_type\n","K80_chk = !nvidia-smi | grep \"Tesla K80\" | wc -l\n","if \"1\" in K80_chk:\n","  print(\"WARNING: found GPU Tesla K80: limited to total length < 1000\")\n","  if \"TF_FORCE_UNIFIED_MEMORY\" in os.environ:\n","    del os.environ[\"TF_FORCE_UNIFIED_MEMORY\"]\n","  if \"XLA_PYTHON_CLIENT_MEM_FRACTION\" in os.environ:\n","    del os.environ[\"XLA_PYTHON_CLIENT_MEM_FRACTION\"]\n","\n","from colabfold.colabfold import plot_protein\n","from pathlib import Path\n","import matplotlib.pyplot as plt\n","\n","\n","# For some reason we need that to get pdbfixer to import\n","if use_amber and '/usr/local/lib/python3.7/site-packages/' not in sys.path:\n","    sys.path.insert(0, '/usr/local/lib/python3.7/site-packages/')\n","\n","def prediction_callback(unrelaxed_protein, length, prediction_result, input_features, type):\n","  fig = plot_protein(unrelaxed_protein, Ls=length, dpi=150)\n","  plt.show()\n","  plt.close()\n","\n","result_dir=\".\"\n","if 'logging_setup' not in globals():\n","    setup_logging(Path(\".\").joinpath(\"log.txt\"))\n","    logging_setup = True\n","\n","queries, is_complex = get_queries(queries_path)\n","model_type = set_model_type(is_complex, model_type)\n","download_alphafold_params(model_type, Path(\".\"))\n","run(\n","    queries=queries,\n","    result_dir=result_dir,\n","    use_templates=use_templates,\n","    custom_template_path=custom_template_path,\n","    use_amber=use_amber,\n","    msa_mode=msa_mode,    \n","    model_type=model_type,\n","    num_models=5,\n","    num_recycles=num_recycles,\n","    model_order=[1, 2, 3, 4, 5],\n","    is_complex=is_complex,\n","    data_dir=Path(\".\"),\n","    keep_existing_results=False,\n","    recompile_padding=1.0,\n","    rank_by=\"auto\",\n","    pair_mode=pair_mode,\n","    stop_at_score=float(100),\n","    prediction_callback=prediction_callback,\n","    dpi=dpi\n",")"],"execution_count":null,"outputs":[]},{"cell_type":"code","metadata":{"id":"KK7X9T44pWb7","cellView":"form"},"source":["#@title Display 3D structure {run: \"auto\"}\n","import py3Dmol\n","import glob\n","import matplotlib.pyplot as plt\n","from colabfold.colabfold import plot_plddt_legend\n","rank_num = 1 #@param [\"1\", \"2\", \"3\", \"4\", \"5\"] {type:\"raw\"}\n","color = \"lDDT\" #@param [\"chain\", \"lDDT\", \"rainbow\"]\n","show_sidechains = False #@param {type:\"boolean\"}\n","show_mainchains = False #@param {type:\"boolean\"}\n","\n","jobname_prefix = \".custom\" if msa_mode == \"custom\" else \"\"\n","if use_amber:\n","  pdb_filename = f\"{jobname}{jobname_prefix}_relaxed_rank_{rank_num}_model_*.pdb\"\n","else:\n","  pdb_filename = f\"{jobname}{jobname_prefix}_unrelaxed_rank_{rank_num}_model_*.pdb\"\n","\n","pdb_file = glob.glob(pdb_filename)\n","\n","def show_pdb(rank_num=1, show_sidechains=False, show_mainchains=False, color=\"lDDT\"):\n","  model_name = f\"rank_{rank_num}\"\n","  view = py3Dmol.view(js='https://3dmol.org/build/3Dmol.js',)\n","  view.addModel(open(pdb_file[0],'r').read(),'pdb')\n","\n","  if color == \"lDDT\":\n","    view.setStyle({'cartoon': {'colorscheme': {'prop':'b','gradient': 'roygb','min':50,'max':90}}})\n","  elif color == \"rainbow\":\n","    view.setStyle({'cartoon': {'color':'spectrum'}})\n","  elif color == \"chain\":\n","    chains = len(queries[0][1]) + 1 if is_complex else 1\n","    for n,chain,color in zip(range(chains),list(\"ABCDEFGH\"),\n","                     [\"lime\",\"cyan\",\"magenta\",\"yellow\",\"salmon\",\"white\",\"blue\",\"orange\"]):\n","      view.setStyle({'chain':chain},{'cartoon': {'color':color}})\n","  if show_sidechains:\n","    BB = ['C','O','N']\n","    view.addStyle({'and':[{'resn':[\"GLY\",\"PRO\"],'invert':True},{'atom':BB,'invert':True}]},\n","                        {'stick':{'colorscheme':f\"WhiteCarbon\",'radius':0.3}})\n","    view.addStyle({'and':[{'resn':\"GLY\"},{'atom':'CA'}]},\n","                        {'sphere':{'colorscheme':f\"WhiteCarbon\",'radius':0.3}})\n","    view.addStyle({'and':[{'resn':\"PRO\"},{'atom':['C','O'],'invert':True}]},\n","                        {'stick':{'colorscheme':f\"WhiteCarbon\",'radius':0.3}})  \n","  if show_mainchains:\n","    BB = ['C','O','N','CA']\n","    view.addStyle({'atom':BB},{'stick':{'colorscheme':f\"WhiteCarbon\",'radius':0.3}})\n","\n","  view.zoomTo()\n","  return view\n","\n","\n","show_pdb(rank_num,show_sidechains, show_mainchains, color).show()\n","if color == \"lDDT\":\n","  plot_plddt_legend().show() "],"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["## # --> Take a screen shot \n","\n","that shows the py3Dmol view and level of confidence in your models. Save it in your `03-colab_fold` folder as `CYPID.png`, where `CYPID` is the identifier of your Cytochrome P450."],"metadata":{"id":"AjlJP6pL-t9m"}},{"cell_type":"markdown","source":["## --> Enter a short answer in this text box\n","\n","Summarize the confidence in your models. What regions in the protein is confidence lower?"],"metadata":{"id":"hHbHnEw0-546"}},{"cell_type":"code","metadata":{"id":"11l8k--10q0C","cellView":"form"},"source":["#@title Plots {run: \"auto\"}\n","from IPython.display import display, HTML\n","import base64\n","from html import escape\n","\n","# see: https://stackoverflow.com/a/53688522\n","def image_to_data_url(filename):\n","  ext = filename.split('.')[-1]\n","  prefix = f'data:image/{ext};base64,'\n","  with open(filename, 'rb') as f:\n","    img = f.read()\n","  return prefix + base64.b64encode(img).decode('utf-8')\n","\n","pae = image_to_data_url(f\"{jobname}{jobname_prefix}_PAE.png\")\n","cov = image_to_data_url(f\"{jobname}{jobname_prefix}_coverage.png\")\n","plddt = image_to_data_url(f\"{jobname}{jobname_prefix}_plddt.png\")\n","display(HTML(f\"\"\"\n","<style>\n","  img {{\n","    float:left;\n","  }}\n","  .full {{\n","    max-width:100%;\n","  }}\n","  .half {{\n","    max-width:50%;\n","  }}\n","  @media (max-width:640px) {{\n","    .half {{\n","      max-width:100%;\n","    }}\n","  }}\n","</style>\n","<div style=\"max-width:90%; padding:2em;\">\n","  <h1>Plots for {escape(jobname)}</h1>\n","  <img src=\"{pae}\" class=\"full\" />\n","  <img src=\"{cov}\" class=\"half\" />\n","  <img src=\"{plddt}\" class=\"half\" />\n","</div>\n","\"\"\"))\n"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["For the Sequence coverage plot, the ColabFold documentation explains that\n","\n","> Number of sequences per position - We want to see at least 30 sequences per position, for best performance, ideally 100 sequences.\n","Predicted lDDT per position - model confidence (out of 100) at each position. The higher the better."],"metadata":{"id":"btDE3T0g-ZL8"}},{"cell_type":"markdown","source":["## --> Enter a short answer in this text box\n","\n","For your CYPs, is their adequate sequence coverage for all positions? If not, in what positions is there low coverage?"],"metadata":{"id":"-0R2hCz--BYM"}},{"cell_type":"code","metadata":{"id":"33g5IIegij5R","cellView":"form"},"source":["#@title Package and download results\n","#@markdown If you are having issues downloading the result archive, try disabling your adblocker and run this cell again. If that fails click on the little folder icon to the left, navigate to file: `jobname.result.zip`, right-click and select \\\"Download\\\" (see [screenshot](https://pbs.twimg.com/media/E6wRW2lWUAEOuoe?format=jpg&name=small)).\n","\n","if msa_mode == \"custom\":\n","  print(\"Don't forget to cite your custom MSA generation method.\")\n","\n","!zip -FSr $jobname\".result.zip\" config.json $jobname*\".json\" $jobname*\".a3m\" $jobname*\"relaxed_rank_\"*\".pdb\" \"cite.bibtex\" $jobname*\".png\"\n","# files.download(f\"{jobname}.result.zip\")\n","\n","if save_to_google_drive == True:\n","  from google.colab import drive\n","  if not os.path.isdir('/content/drive'):\n","    drive.mount('/content/drive')\n","\n","  import os\n","  class_dir = os.path.join('/content','drive','MyDrive','Chem456-2022F')\n","  if not os.path.isdir(class_dir):\n","    raise ValueError('Class directory not set up')\n","  else:\n","    os.chdir(class_dir)\n","\n","  labs_dir = os.path.join(class_dir,'labs')\n","  if not os.path.isdir(labs_dir):\n","    raise ValueError('Labs directory not set up')\n","  else:\n","    os.chdir(labs_dir)\n","\n","  os.chdir(labs_dir)\n","  colab_fold_dir = os.path.join(labs_dir,'03-colab_fold')\n","  if not os.path.isdir(colab_fold_dir):\n","    os.mkdir('03-colab_fold')\n"," \n","  import shutil\n","  shutil.copy(os.path.join('/content',    f\"{jobname}.result.zip\"),\n","              os.path.join(colab_fold_dir,f\"{jobname}.result.zip\"))\n","  print(f'Saved {jobname}.result.zip to {colab_fold_dir}')\n","  os.chdir('/content')\n","\n","  # uploaded = drive.CreateFile({'title': f\"{jobname}.result.zip\"})\n","  # uploaded.SetContentFile(f\"{jobname}.result.zip\")\n","  # uploaded.Upload()\n","  # print(f\"Uploaded {jobname}.result.zip to Google Drive with ID {uploaded.get('id')}\")"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"UGUBLzB3C6WN","pycharm":{"name":"#%% md\n"}},"source":["# Instructions <a name=\"Instructions\"></a>\n","**Quick start**\n","1. Paste your protein sequence(s) in the input field.\n","2. Press \"Runtime\" -> \"Run all\".\n","3. The pipeline consists of 5 steps. The currently running step is indicated by a circle with a stop sign next to it.\n","\n","**Result zip file contents**\n","\n","1. PDB formatted structures sorted by avg. pLDDT and complexes are sorted by pTMscore. (unrelaxed and relaxed if `use_amber` is enabled).\n","2. Plots of the model quality.\n","3. Plots of the MSA coverage.\n","4. Parameter log file.\n","5. A3M formatted input MSA.\n","6. A `predicted_aligned_error_v1.json` using [AlphaFold-DB's format](https://alphafold.ebi.ac.uk/faq#faq-7) and a `scores.json` for each model which contains an array (list of lists) for PAE, a list with the average pLDDT and the pTMscore.\n","7. BibTeX file with citations for all used tools and databases.\n","\n","At the end of the job a download modal box will pop up with a `jobname.result.zip` file. Additionally, if the `save_to_google_drive` option was selected, the `jobname.result.zip` will be uploaded to your Google Drive.\n","\n","**MSA generation for complexes**\n","\n","For the complex prediction we use unpaired and paired MSAs. Unpaired MSA is generated the same way as for the protein structures prediction by searching the UniRef100 and environmental sequences three iterations each.\n","\n","The paired MSA is generated by searching the UniRef100 database and pairing the best hits sharing the same NCBI taxonomic identifier (=species or sub-species). We only pair sequences if all of the query sequences are present for the respective taxonomic identifier.\n","\n","**Using a custom MSA as input**\n","\n","To predict the structure with a custom MSA (A3M formatted): (1) Change the `msa_mode`: to \"custom\", (2) Wait for an upload box to appear at the end of the \"MSA options ...\" box. Upload your A3M. The first fasta entry of the A3M must be the query sequence without gaps. \n","\n","It is also possilbe to proide custom MSAs for complex predictions. Read more about the format [here](https://github.com/sokrypton/ColabFold/issues/76).\n","\n","As an alternative for MSA generation the [HHblits Toolkit server](https://toolkit.tuebingen.mpg.de/tools/hhblits) can be used. After submitting your query, click \"Query Template MSA\" -> \"Download Full A3M\". Download the A3M file and upload it in this notebook.\n","\n","**Using custom templates** <a name=\"custom_templates\"></a>\n","\n","To predict the structure with a custom template (PDB or mmCIF formatted): (1) change the `template_mode` to \"custom\" in the execute cell and (2) wait for an upload box to appear at the end of the \"Input Protein\" box. Select and upload your templates (multiple choices are possible).\n","\n","* Templates must follow the four letter PDB naming.\n","\n","* Templates in mmCIF format must contain `_entity_poly_seq`. An error is thrown if this field is not present. The field `_pdbx_audit_revision_history.revision_date` is automatically generated if it is not present.\n","\n","* Templates in PDB format are automatically converted to the mmCIF format. `_entity_poly_seq` and `_pdbx_audit_revision_history.revision_date` are automatically generated.\n","\n","If you encounter problems, please report them to this [issue](https://github.com/sokrypton/ColabFold/issues/177).\n","\n","**Comparison to the full AlphaFold2 and Alphafold2 colab**\n","\n","This notebook replaces the homology detection and MSA pairing of AlphaFold2 with MMseqs2. For a comparison against the [AlphaFold2 Colab](https://colab.research.google.com/github/deepmind/alphafold/blob/main/notebooks/AlphaFold.ipynb) and the full [AlphaFold2](https://github.com/deepmind/alphafold) system read our [preprint](https://www.biorxiv.org/content/10.1101/2021.08.15.456425v1). \n","\n","**Troubleshooting**\n","* Check that the runtime type is set to GPU at \"Runtime\" -> \"Change runtime type\".\n","* Try to restart the session \"Runtime\" -> \"Factory reset runtime\".\n","* Check your input sequence.\n","\n","**Known issues**\n","* Google Colab assigns different types of GPUs with varying amount of memory. Some might not have enough memory to predict the structure for a long sequence.\n","* Your browser can block the pop-up for downloading the result file. You can choose the `save_to_google_drive` option to upload to Google Drive instead or manually download the result file: Click on the little folder icon to the left, navigate to file: `jobname.result.zip`, right-click and select \\\"Download\\\" (see [screenshot](https://pbs.twimg.com/media/E6wRW2lWUAEOuoe?format=jpg&name=small)).\n","\n","**Limitations**\n","* Computing resources: Our MMseqs2 API can handle ~20-50k requests per day.\n","* MSAs: MMseqs2 is very precise and sensitive but might find less hits compared to HHblits/HMMer searched against BFD or MGnify.\n","* We recommend to additionally use the full [AlphaFold2 pipeline](https://github.com/deepmind/alphafold).\n","\n","**Description of the plots**\n","*   **Number of sequences per position** - We want to see at least 30 sequences per position, for best performance, ideally 100 sequences.\n","*   **Predicted lDDT per position** - model confidence (out of 100) at each position. The higher the better.\n","*   **Predicted Alignment Error** - For homooligomers, this could be a useful metric to assess how confident the model is about the interface. The lower the better.\n","\n","**Bugs**\n","- If you encounter any bugs, please report the issue to https://github.com/sokrypton/ColabFold/issues\n","\n","**License**\n","\n","The source code of ColabFold is licensed under [MIT](https://raw.githubusercontent.com/sokrypton/ColabFold/main/LICENSE). Additionally, this notebook uses the AlphaFold2 source code and its parameters licensed under [Apache 2.0](https://raw.githubusercontent.com/deepmind/alphafold/main/LICENSE) and [CC BY 4.0](https://creativecommons.org/licenses/by-sa/4.0/) respectively. Read more about the AlphaFold license [here](https://github.com/deepmind/alphafold).\n","\n","**Acknowledgments**\n","- We thank the AlphaFold team for developing an excellent model and open sourcing the software. \n","\n","- [KOBIC](https://kobic.re.kr) and [SÃ¶ding Lab](https://www.mpinat.mpg.de/soeding) for providing the computational resources for the MMseqs2 MSA server.\n","\n","- Richard Evans for helping to benchmark the ColabFold's Alphafold-multimer support.\n","\n","- [David Koes](https://github.com/dkoes) for his awesome [py3Dmol](https://3dmol.csb.pitt.edu/) plugin, without whom these notebooks would be quite boring!\n","\n","- Do-Yoon Kim for creating the ColabFold logo.\n","\n","- A colab by Sergey Ovchinnikov ([@sokrypton](https://twitter.com/sokrypton)), Milot Mirdita ([@milot_mirdita](https://twitter.com/milot_mirdita)) and Martin Steinegger ([@thesteinegger](https://twitter.com/thesteinegger)).\n"]},{"cell_type":"markdown","source":["# Part II. Heme placement"],"metadata":{"id":"Uy7Vgzrry6dB"}},{"cell_type":"markdown","source":["If you run Part II separately from Pa\n","rt I, you will need to upload your results from Part I to the Colab instance and decompress it."],"metadata":{"id":"wcOgUXj9MW1l"}},{"cell_type":"code","source":["# !unzip {jobname}.result.zip"],"metadata":{"id":"giR2CCV-L27e"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["## Compare models from ColabFold and AlphaFold\n","\n","Let's compare the homology model from ColabFold to the model from AlphaFold. The difference between the two methods is the MSA. You should upload the AlphaFold model to your Google Drive `MyDrive/Chem456-2022F/exercises/03-colab_fold` or add a link to it there."],"metadata":{"id":"dlmXGPTmRGBw"}},{"cell_type":"code","source":["# Points to the PDB file from AlphaFold\n","colab_fold_dir = os.path.join('/content','drive','MyDrive','Chem456-2022F', \\\n","                                'exercise','03-colab_fold')\n","\n","AlphaFold_PDBFN = os.path.join(colab_fold_dir, f'{basejobname}-m1.pdb')\n","\n","# Point to the best model from ColabFold\n","for FN in os.listdir('/content'):\n","  if FN.find('unrelaxed_rank_1')>-1 and FN.endswith('pdb'):\n","    ColabFold_PDBFN = os.path.join('/content',FN)\n","\n","print(f'Best ColabFold model is {ColabFold_PDBFN}')\n","\n","Aligned_ColabFold_PDBFN = os.path.join(colab_fold_dir, \\\n","                                       f'{basejobname}_ColabFold.pdb')"],"metadata":{"id":"PeIpFyO4ZWfe"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["# Align the ColabFold to the AlphaFold model\n","print(f'Loading AlphaFold model from {AlphaFold_PDBFN}')\n","AlphaFold_PDB = prody.parsePDB(AlphaFold_PDBFN)\n","print(f'Loading ColabFold model from {ColabFold_PDBFN}')\n","ColabFold_PDB = prody.parsePDB(ColabFold_PDBFN)\n","\n","AlphaFold_PDB_prot = AlphaFold_PDB.select('protein')\n","ColabFold_PDB_prot = ColabFold_PDB.select('protein')\n","\n","map_AC, map_CA, seqid, overlap = \\\n","  prody.matchChains(AlphaFold_PDB_prot, ColabFold_PDB_prot)[0]\n","transformation = prody.calcTransformation(map_CA, map_AC)\n","\n","prody.applyTransformation(transformation, ColabFold_PDB)\n","prody.writePDB(Aligned_ColabFold_PDBFN, ColabFold_PDB)"],"metadata":{"id":"Hj5FqMCUa1dz"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["view = py3Dmol.view()\n","view.addModel(open(AlphaFold_PDBFN, 'r').read(),'pdb')\n","view.addModel(open(Aligned_ColabFold_PDBFN, 'r').read(),'pdb')\n","view.setBackgroundColor('white')\n","view.setStyle({'model':0}, {'cartoon': {'color':'purple'}})\n","view.setStyle({'model':1}, {'cartoon': {'color':'yellow'}})\n","# view.addStyle({'model':0}, {'stick': {'colorscheme':'purpleCarbon'}})\n","# view.addStyle({'model':1}, {'stick': {'colorscheme':'yellowCarbon'}})\n","view.zoomTo()\n","view.show()"],"metadata":{"id":"FKCzG9EpVRku"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["## # --> Take a screen shot \n","\n","that shows the py3Dmol view of the aligned models. Save it in your `03-colab_fold` folder as `CYPID_model_comparison.png`, where `CYPID` is the identifier of your Cytochrome P450."],"metadata":{"id":"vKdG54ftj0Wz"}},{"cell_type":"markdown","source":["## # --> Enter a short answer in this text box\n","\n","For each of your CYPs, compare and contrast the models from ColabFold and AlphaFold."],"metadata":{"id":"GNuSBw98kZnG"}},{"cell_type":"markdown","source":["## Align the closest template to the AlphaFold model"],"metadata":{"id":"XgD3c_QmlKVo"}},{"cell_type":"code","source":["# Template file name\n","Aligned_template_PDBFN = os.path.join(colab_fold_dir, f'{basejobname}_template.pdb')\n","\n","Aligned_heme_PDBFN = os.path.join(colab_fold_dir, f'{basejobname}_heme.pdb')"],"metadata":{"id":"nJfKPXTEqvmi"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["# Identify the closest template from the ColabFold data files\n","os.chdir('/content')\n","import json\n","with open(f\"{jobname}_template_domain_names.json\",'r') as F:\n","  templates = json.loads(F.read())\n","(template_PDBID, template_chain) = templates['A'][0].split('_')"],"metadata":{"id":"KynXlzo-vOQE"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["# Align the template to the AlphaFold model\n","AlphaFold_PDB = prody.parsePDB(AlphaFold_PDBFN)\n","template_PDB = prody.parsePDB(template_PDBID)\n","\n","AlphaFold_PDB_prot = AlphaFold_PDB.select('protein')\n","template_PDB_prot = template_PDB.select('protein')\n","\n","map_AT, map_TA, seqid, overlap = \\\n","  prody.matchChains(AlphaFold_PDB_prot, template_PDB_prot, seqid=10, overlap=10)[0]\n","transformation = prody.calcTransformation(map_TA, map_AT)\n","\n","prody.applyTransformation(transformation, template_PDB)\n","prody.writePDB(Aligned_template_PDBFN, template_PDB)"],"metadata":{"id":"c7CPk--FTE3S"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["print(f'The closest template is {template_PDBID}.')\n","print('These are the unique residues in the template:')\n","print(' '.join(set(list(template_PDB.getResnames()))))"],"metadata":{"id":"tSNF_TGj0t_y"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["## # --> Enter a short answer in this text box\n","\n","For each of your CYPs, which residues are not standard amino acids? Three-letter codes are widely available [online](https://www.fao.org/3/Y2775E/y2775e0e.htm). You may also consult the [PDB](https://www.rcsb.org/)."],"metadata":{"id":"ZItg-ArP2cpd"}},{"cell_type":"code","source":["# --> Enter the residue name of the heme. It is probably 'HEM'\n","heme_resname = 'HEM'"],"metadata":{"id":"oPRCpWb86vTl"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["The code below should show the AlphaFold model, the closest template, and the heme. If not, make sure that the style has the correct residue name for your template."],"metadata":{"id":"s6ooOebq5Lqc"}},{"cell_type":"code","source":["template_PDB_heme = template_PDB.select(f'resname {heme_resname}')\n","prody.writePDB(Aligned_heme_PDBFN, template_PDB_heme)"],"metadata":{"id":"vbUgXwRb55gr"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["view = py3Dmol.view()\n","view.addModel(open(AlphaFold_PDBFN, 'r').read(),'pdb')\n","view.addModel(open(Aligned_template_PDBFN, 'r').read(),'pdb')\n","view.setBackgroundColor('white')\n","view.setStyle({'model':0}, {'cartoon': {'color':'purple'}})\n","view.setStyle({'model':1}, {'cartoon': {'color':'yellow'}})\n","view.addStyle({'resn':'HEM'}, {'sphere': {'colorscheme':'yellowCarbon'}})\n","# view.addStyle({'model':0}, {'stick': {'colorscheme':'purpleCarbon'}})\n","# view.addStyle({'model':1}, {'stick': {'colorscheme':'yellowCarbon'}})\n","view.zoomTo()\n","view.show()"],"metadata":{"id":"ZT41hy_psEfw"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["## # --> Take a screen shot \n","\n","that shows the py3Dmol view of the aligned models and shows the heme. Save it in your `03-colab_fold` folder as `CYPID_with_heme.png`, where `CYPID` is the identifier of your Cytochrome P450."],"metadata":{"id":"uyTblfDR5i4L"}},{"cell_type":"code","source":["# This code should join the AlphaFold model with the heme from the template\n","AlphaFoldHeme_PDBFN = os.path.join(colab_fold_dir, f'{basejobname}_with_heme.pdb')\n","!cat {AlphaFold_PDBFN} {Aligned_heme_PDBFN} > {AlphaFoldHeme_PDBFN}"],"metadata":{"id":"LqaFOECO6HBN"},"execution_count":null,"outputs":[]}]}